In [ ]:
!mkdir -p genome-annotation-feature-overlap
%cd genome-annotation-feature-overlap

!mkdir -p data/raw data/processed results/figures scripts

/content/genome-annotation-feature-overlap


In [ ]:
!ls

data  results  scripts


In [ ]:
!ls data

processed  raw


In [ ]:
!apt-get update -qq
!apt-get install -y bedtools samtools

W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  libhts3 libhtscodecs2
Suggested packages:
  cwltool
The following NEW packages will be installed:
  bedtools libhts3 libhtscodecs2 samtools
0 upgraded, 4 newly installed, 0 to remove and 49 not upgraded.
Need to get 1,525 kB of archives.
After this operation, 3,818 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 bedtools amd64 2.30.0+dfsg-2ubuntu0.1 [563 kB]
Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]
Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]
Get:4 http://archive.ubuntu.com/ubuntu 

In [ ]:
!bedtools --version
!samtools --version

bedtools v2.30.0
samtools 1.13
Using htslib 1.13+ds
Copyright (C) 2021 Genome Research Ltd.

Samtools compilation details:
    Features:       build=configure curses=yes 
    CC:             gcc
    CPPFLAGS:       -frelease  -Wdate-time -D_FORTIFY_SOURCE=2
    CFLAGS:         -g -O2 -ffile-prefix-map=�BUILDPATH�=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security
    LDFLAGS:        -Wl,-Bsymbolic-functions -flto=auto -Wl,-z,relro -Wl,-z,now
    HTSDIR:         
    LIBS:           
    CURSES_LIB:     -lcurses

HTSlib compilation details:
    Features:       build=configure plugins=yes, plugin-path=/usr/local/lib/htslib:/usr/local/libexec/htslib:: libcurl=yes S3=yes GCS=yes libdeflate=yes lzma=yes bzip2=yes htscodecs=1.1.1
    CC:             gcc
    CPPFLAGS:       -I. -DSAMTOOLS=1 -Wdate-time -D_FORTIFY_SOURCE=2
    CFLAGS:         -g -O2 -ffile-prefix-map=/build/htslib-TQtOKr/htslib-1.13+ds=. -flto=auto -ffat-lto-objects -fstack-protector-stron

In [ ]:
%cd /content/genome-annotation-feature-overlap/data/raw
!wget -o annotation.gff3.gz
"https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-63/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.63.gff3.gz"
!wget -o genome.fa.gz
"https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-63/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz"

!gunzip annotation.gff3.gz
!gunzip genome.fa.gz
%cd /content/genome-annotation-feature-overlap/

/content/genome-annotation-feature-overlap/data/raw
wget: missing URL
Usage: wget [OPTION]... [URL]...

Try `wget --help' for more options.
wget: missing URL
Usage: wget [OPTION]... [URL]...

Try `wget --help' for more options.

gzip: annotation.gff3.gz: unexpected end of file

gzip: genome.fa.gz: unexpected end of file
/content/genome-annotation-feature-overlap


In [ ]:
!ls -lh data/raw

total 0
-rw-r--r-- 1 root root 0 Jun 22 21:50 annotation.gff3.gz
-rw-r--r-- 1 root root 0 Jun 22 21:50 genome.fa.gz


In [ ]:
!pwd

/content/genome-annotation-feature-overlap


In [ ]:
!ls

data  results  scripts


In [ ]:

%cd /content/genome-annotation-feature-overlap/data/raw

!wget -O annotation.gff3.gz "https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-63/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.63.gff3.gz"

!wget -O genome.fa.gz "https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-63/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz"

/content/genome-annotation-feature-overlap/data/raw
--2026-06-22 22:34:10--  https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-63/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.63.gff3.gz
Resolving ftp.ensemblgenomes.ebi.ac.uk (ftp.ensemblgenomes.ebi.ac.uk)... 193.62.193.161
Connecting to ftp.ensemblgenomes.ebi.ac.uk (ftp.ensemblgenomes.ebi.ac.uk)|193.62.193.161|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 9571450 (9.1M) [application/x-gzip]
Saving to: ‘annotation.gff3.gz’

annotation.gff3.gz  100%[===================>]   9.13M  10.5MB/s    in 0.9s    

2026-06-22 22:34:12 (10.5 MB/s) - ‘annotation.gff3.gz’ saved [9571450/9571450]

--2026-06-22 22:34:12--  https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-63/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz
Resolving ftp.ensemblgenomes.ebi.ac.uk (ftp.ensemblgenomes.ebi.ac.uk)... 193.62.193.161
Connecting to ftp.ensemblgenomes.ebi.ac.uk (ftp.ensemblgenomes.ebi.a

In [ ]:
!ls -lh

total 44M
-rw-r--r-- 1 root root 9.2M Apr  6 02:50 annotation.gff3.gz
-rw-r--r-- 1 root root  35M Apr  3 06:19 genome.fa.gz


In [ ]:
!ls -lh

total 44M
-rw-r--r-- 1 root root 9.2M Apr  6 02:50 annotation.gff3.gz
-rw-r--r-- 1 root root  35M Apr  3 06:19 genome.fa.gz


In [ ]:
!gunzip annotation.gff3.gz
!gunzip genome.fa.gz

gzip: annotation.gff3.gz: No such file or directory
gzip: genome.fa.gz: No such file or directory


In [ ]:
!ls -lh

total 224M
-rw-r--r-- 1 root root 108M Apr  6 02:50 annotation.gff3
-rw-r--r-- 1 root root 117M Apr  3 06:19 genome.fa


In [ ]:
%cd /content/genome-annotation-feature-overlap

/content/genome-annotation-feature-overlap


In [ ]:
!pwd

/content/genome-annotation-feature-overlap


In [ ]:
!ls -lh data/raw

total 224M
-rw-r--r-- 1 root root 108M Apr  6 02:50 annotation.gff3
-rw-r--r-- 1 root root 117M Apr  3 06:19 genome.fa


In [ ]:
!head data/raw/annotation.gff3

##gff-version 3
##sequence-region   1 1 30427671
##sequence-region   2 1 19698289
##sequence-region   3 1 23459830
##sequence-region   4 1 18585056
##sequence-region   5 1 26975502
##sequence-region   Mt 1 366924
##sequence-region   Pt 1 154478
#!genome-build The Arabidopsis Information Resource TAIR10
#!genome-version TAIR10


In [ ]:
!grep -v "^#" data/raw/annotation.gff3 | head

1	TAIR10	chromosome	1	30427671	.	.	.	ID=chromosome:1;Alias=Chr1,CP002684.1,NC_003070.9
1	araport11	gene	3631	5899	.	+	.	ID=gene:AT1G01010;Name=NAC001;biotype=protein_coding;description=NAC domain containing protein 1 [Source:NCBI gene (formerly Entrezgene)%3BAcc:839580];gene_id=AT1G01010;logic_name=araport11
1	araport11	mRNA	3631	5899	.	+	.	ID=transcript:AT1G01010.1;Parent=gene:AT1G01010;Name=NAC001-201;biotype=protein_coding;tag=Ensembl_canonical;transcript_id=AT1G01010.1
1	araport11	five_prime_UTR	3631	3759	.	+	.	Parent=transcript:AT1G01010.1
1	araport11	exon	3631	3913	.	+	.	Parent=transcript:AT1G01010.1;Name=AT1G01010.1.exon1;constitutive=1;ensembl_end_phase=1;ensembl_phase=-1;exon_id=AT1G01010.1.exon1;rank=1
1	araport11	CDS	3760	3913	.	+	0	ID=CDS:AT1G01010.1;Parent=transcript:AT1G01010.1;protein_id=AT1G01010.1
1	araport11	exon	3996	4276	.	+	.	Parent=transcript:AT1G01010.1;Name=AT1G01010.1.exon2;constitutive=1;ensembl_end_phase=0;ensembl_phase=1;exon_id=AT1G01010.1.exon2;rank=2
1	ar

In [ ]:
!grep -v "^#" data/raw/annotation.gff3 | cut -f3 | sort | uniq -c | sort -nr | head

 313952 exon
 286067 CDS
  56384 five_prime_UTR
  48359 mRNA
  48308 three_prime_UTR
  27655 gene
   5178 ncRNA_gene
   3879 lnc_RNA
    689 tRNA
    377 ncRNA


## Annotation File Inspection

The genome annotation file was inspected to identify available genomic feature types. Comment lines were removed, and the third column of the GFF3 file was used to summarize feature categories such as genes, exons, CDS regions, transcripts, and UTRs.

In [ ]:
%%bash
cd /content/genome-annotation-feature-overlap

awk 'BEGIN{OFS="\t"}
!/^#/ && $3=="gene" {
  id=".";
  n=split($9,a,";");
  for(i=1;i<=n;i++){
    split(a[i],b,"=");
    if(b[1]=="ID"){
      id=b[2];
    }
  }
  gsub(/^gene:/,"",id);
  print $1,$4-1,$5,id,$6,$7
}' data/raw/annotation.gff3 > data/processed/genes.bed

In [ ]:
!head data/processed/genes.bed

1	3630	5899	AT1G01010	.	+
1	6787	9130	AT1G01020	.	-
1	11648	13714	AT1G01030	.	-
1	23120	31227	AT1G01040	.	+
1	31169	33171	AT1G01050	.	-
1	33364	37871	AT1G01060	.	-
1	38443	41017	AT1G01070	.	-
1	44969	47059	AT1G01080	.	-
1	47233	49304	AT1G01090	.	-
1	49908	51210	AT1G01100	.	-


In [ ]:
!wc -l data/processed/genes.bed

27655 data/processed/genes.bed


In [ ]:
%%bash
cd /content/genome-annotation-feature-overlap

awk 'BEGIN{OFS="\t"}
!/^#/ && $3=="exon" {
  id=".";
  n=split($9,a,";");
  for(i=1;i<=n;i++){
    split(a[i],b,"=");
    if(b[1]=="ID"){
      id=b[2];
    }
  }
  gsub(/^exon:/,"",id);
  print $1,$4-1,$5,id,$6,$7
}' data/raw/annotation.gff3 > data/processed/exons.bed

In [ ]:
!head data/processed/exons.bed
!wc -l data/processed/exons.bed

1	3630	3913	.	.	+
1	3995	4276	.	.	+
1	4485	4605	.	.	+
1	4705	5095	.	.	+
1	5173	5326	.	.	+
1	5438	5899	.	.	+
1	6787	7069	.	.	-
1	7156	7450	.	.	-
1	7563	7649	.	.	-
1	7761	7835	.	.	-
313952 data/processed/exons.bed


In [ ]:
%%bash
cd /content/genome-annotation-feature-overlap

awk 'BEGIN{OFS="\t"}
!/^#/ && $3=="CDS" {
  id=".";
  n=split($9,a,";");
  for(i=1;i<=n;i++){
    split(a[i],b,"=");
    if(b[1]=="ID"){
      id=b[2];
    }
  }
  gsub(/^CDS:/,"",id);
  print $1,$4-1,$5,id,$6,$7
}' data/raw/annotation.gff3 > data/processed/cds.bed

In [ ]:
!head data/processed/cds.bed
!wc -l data/processed/cds.bed

1	3759	3913	AT1G01010.1	.	+
1	3995	4276	AT1G01010.1	.	+
1	4485	4605	AT1G01010.1	.	+
1	4705	5095	AT1G01010.1	.	+
1	5173	5326	AT1G01010.1	.	+
1	5438	5630	AT1G01010.1	.	+
1	7314	7450	AT1G01020.2	.	-
1	7563	7649	AT1G01020.2	.	-
1	7761	7835	AT1G01020.2	.	-
1	7941	7987	AT1G01020.2	.	-
286067 data/processed/cds.bed


In [ ]:
!ls -lh data/processed

total 18M
-rw-r--r-- 1 root root 9.6M Jun 22 23:09 cds.bed
-rw-r--r-- 1 root root 7.5M Jun 22 23:09 exons.bed
-rw-r--r-- 1 root root 892K Jun 22 23:07 genes.bed


In [ ]:
!wc -l data/processed/genes.bed
!wc -l data/processed/exons.bed
!wc -l data/processed/cds.bed

27655 data/processed/genes.bed
313952 data/processed/exons.bed
286067 data/processed/cds.bed


In [ ]:
!head data/processed/genes.bed

1	3630	5899	AT1G01010	.	+
1	6787	9130	AT1G01020	.	-
1	11648	13714	AT1G01030	.	-
1	23120	31227	AT1G01040	.	+
1	31169	33171	AT1G01050	.	-
1	33364	37871	AT1G01060	.	-
1	38443	41017	AT1G01070	.	-
1	44969	47059	AT1G01080	.	-
1	47233	49304	AT1G01090	.	-
1	49908	51210	AT1G01100	.	-


In [ ]:
!head data/processed/exons.bed

1	3630	3913	.	.	+
1	3995	4276	.	.	+
1	4485	4605	.	.	+
1	4705	5095	.	.	+
1	5173	5326	.	.	+
1	5438	5899	.	.	+
1	6787	7069	.	.	-
1	7156	7450	.	.	-
1	7563	7649	.	.	-
1	7761	7835	.	.	-


In [ ]:
!head data/processed/cds.bed

1	3759	3913	AT1G01010.1	.	+
1	3995	4276	AT1G01010.1	.	+
1	4485	4605	AT1G01010.1	.	+
1	4705	5095	AT1G01010.1	.	+
1	5173	5326	AT1G01010.1	.	+
1	5438	5630	AT1G01010.1	.	+
1	7314	7450	AT1G01020.2	.	-
1	7563	7649	AT1G01020.2	.	-
1	7761	7835	AT1G01020.2	.	-
1	7941	7987	AT1G01020.2	.	-


In [ ]:
!samtools faidx data/raw/genome.fa

In [ ]:
!cut -f1,2 data/raw/genome.fa.fai > data/processed/genome_file.txt

In [ ]:
!head data/processed/genome_file.txt

1	30427671
2	19698289
3	23459830
4	18585056
5	26975502
Mt	366924
Pt	154478


In [ ]:
!bedtools flank \
  -i data/processed/genes.bed \
  -g data/processed/genome_file.txt \
  -l 1000 \
  -r 0 \
  -s > data/processed/promoters_1kb.bed

In [ ]:
!head data/processed/promoters_1kb.bed

1	2630	3630	AT1G01010	.	+
1	9130	10130	AT1G01020	.	-
1	13714	14714	AT1G01030	.	-
1	22120	23120	AT1G01040	.	+
1	33171	34171	AT1G01050	.	-
1	37871	38871	AT1G01060	.	-
1	41017	42017	AT1G01070	.	-
1	47059	48059	AT1G01080	.	-
1	49304	50304	AT1G01090	.	-
1	51210	52210	AT1G01100	.	-


In [ ]:
!wc -l data/processed/promoters_1kb.bed

27655 data/processed/promoters_1kb.bed


## Feature Extraction

The genome annotation file was parsed to extract genes, exons, and CDS regions. These features were converted from GFF3 format into BED format for downstream interval-based analysis using BEDTools.

Extracted feature counts:

| Feature Type | Count |
|---|---:|
| Genes | 27,655 |
| Exons | 313,952 |
| CDS regions | 286,067 |

In [ ]:
%cd /content/genome-annotation-feature-overlap

/content/genome-annotation-feature-overlap


In [ ]:
!pwd

/content/genome-annotation-feature-overlap


In [ ]:
!ls -lh data/raw

total 224M
-rw-r--r-- 1 root root 108M Apr  6 02:50 annotation.gff3
-rw-r--r-- 1 root root 117M Apr  3 06:19 genome.fa
-rw-r--r-- 1 root root  176 Jun 22 23:14 genome.fa.fai


In [ ]:
!ls -lh data/processed

total 19M
-rw-r--r-- 1 root root 9.6M Jun 22 23:09 cds.bed
-rw-r--r-- 1 root root 7.5M Jun 22 23:09 exons.bed
-rw-r--r-- 1 root root 892K Jun 22 23:07 genes.bed
-rw-r--r-- 1 root root   75 Jun 22 23:14 genome_file.txt
-rw-r--r-- 1 root root 892K Jun 22 23:15 promoters_1kb.bed


In [ ]:
!samtools faidx data/raw/genome.fa

/bin/bash: line 1: samtools: command not found


In [ ]:
!apt-get update -qq
!apt-get install -y samtools bedtools

W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)
Reading package lists... Done
Building dependency tree... Done
Reading state information... Done
The following additional packages will be installed:
  libhts3 libhtscodecs2
Suggested packages:
  cwltool
The following NEW packages will be installed:
  bedtools libhts3 libhtscodecs2 samtools
0 upgraded, 4 newly installed, 0 to remove and 49 not upgraded.
Need to get 1,525 kB of archives.
After this operation, 3,818 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy-updates/universe amd64 bedtools amd64 2.30.0+dfsg-2ubuntu0.1 [563 kB]
Get:2 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhtscodecs2 amd64 1.1.1-3 [53.2 kB]
Get:3 http://archive.ubuntu.com/ubuntu jammy/universe amd64 libhts3 amd64 1.13+ds-2build1 [390 kB]
Get:4 http://archive.ubuntu.com/ubuntu 

In [ ]:
!samtools --version
!bedtools --version

samtools 1.13
Using htslib 1.13+ds
Copyright (C) 2021 Genome Research Ltd.

Samtools compilation details:
    Features:       build=configure curses=yes 
    CC:             gcc
    CPPFLAGS:       -frelease  -Wdate-time -D_FORTIFY_SOURCE=2
    CFLAGS:         -g -O2 -ffile-prefix-map=�BUILDPATH�=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werror=format-security
    LDFLAGS:        -Wl,-Bsymbolic-functions -flto=auto -Wl,-z,relro -Wl,-z,now
    HTSDIR:         
    LIBS:           
    CURSES_LIB:     -lcurses

HTSlib compilation details:
    Features:       build=configure plugins=yes, plugin-path=/usr/local/lib/htslib:/usr/local/libexec/htslib:: libcurl=yes S3=yes GCS=yes libdeflate=yes lzma=yes bzip2=yes htscodecs=1.1.1
    CC:             gcc
    CPPFLAGS:       -I. -DSAMTOOLS=1 -Wdate-time -D_FORTIFY_SOURCE=2
    CFLAGS:         -g -O2 -ffile-prefix-map=/build/htslib-TQtOKr/htslib-1.13+ds=. -flto=auto -ffat-lto-objects -fstack-protector-strong -Wformat -Werro

In [ ]:
%cd /content/genome-annotation-feature-overlap
!pwd

[Errno 2] No such file or directory: '/content/genome-annotation-feature-overlap'
/content
/content


In [ ]:
!ls -lh /content

total 4.0K
drwxr-xr-x 1 root root 4.0K Jun  4 13:32 sample_data


In [ ]:
%%bash

cd /content

# Create project folders
mkdir -p genome-annotation-feature-overlap
cd genome-annotation-feature-overlap
mkdir -p data/raw data/processed results/figures scripts

# Install required tools
apt-get update -qq
apt-get install -y samtools bedtools

# Download raw data
cd data/raw

wget -O annotation.gff3.gz "https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-63/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.63.gff3.gz"

wget -O genome.fa.gz "https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-63/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz"

gunzip -f annotation.gff3.gz
gunzip -f genome.fa.gz

cd /content/genome-annotation-feature-overlap

# Extract genes
awk 'BEGIN{OFS="\t"}
!/^#/ && $3=="gene" {
  id=".";
  n=split($9,a,";");
  for(i=1;i<=n;i++){
    split(a[i],b,"=");
    if(b[1]=="ID"){
      id=b[2];
    }
  }
  gsub(/^gene:/,"",id);
  print $1,$4-1,$5,id,$6,$7
}' data/raw/annotation.gff3 > data/processed/genes.bed

# Extract exons
awk 'BEGIN{OFS="\t"}
!/^#/ && $3=="exon" {
  id=".";
  n=split($9,a,";");
  for(i=1;i<=n;i++){
    split(a[i],b,"=");
    if(b[1]=="ID"){
      id=b[2];
    }
  }
  gsub(/^exon:/,"",id);
  print $1,$4-1,$5,id,$6,$7
}' data/raw/annotation.gff3 > data/processed/exons.bed

# Extract CDS
awk 'BEGIN{OFS="\t"}
!/^#/ && $3=="CDS" {
  id=".";
  n=split($9,a,";");
  for(i=1;i<=n;i++){
    split(a[i],b,"=");
    if(b[1]=="ID"){
      id=b[2];
    }
  }
  gsub(/^CDS:/,"",id);
  print $1,$4-1,$5,id,$6,$7
}' data/raw/annotation.gff3 > data/processed/cds.bed

# Index genome FASTA
samtools faidx data/raw/genome.fa

# Create genome size file
cut -f1,2 data/raw/genome.fa.fai > data/processed/genome_file.txt

# Create 1 kb promoter regions
bedtools flank \
  -i data/processed/genes.bed \
  -g data/processed/genome_file.txt \
  -l 1000 \
  -r 0 \
  -s > data/processed/promoters_1kb.bed

echo "Recovery completed successfully."

Reading package lists...
Building dependency tree...
Reading state information...
samtools is already the newest version (1.13-4).
bedtools is already the newest version (2.30.0+dfsg-2ubuntu0.1).
0 upgraded, 0 newly installed, 0 to remove and 49 not upgraded.
Recovery completed successfully.


W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)
--2026-06-23 05:16:38--  https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-63/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.63.gff3.gz
Resolving ftp.ensemblgenomes.ebi.ac.uk (ftp.ensemblgenomes.ebi.ac.uk)... 193.62.193.161
Connecting to ftp.ensemblgenomes.ebi.ac.uk (ftp.ensemblgenomes.ebi.ac.uk)|193.62.193.161|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 9571450 (9.1M) [application/x-gzip]
Saving to: ‘annotation.gff3.gz’

     0K .......... .......... .......... .......... ..........  0%  254K 37s
    50K .......... .......... .......... .......... ..........  1% 54.5M 18s
   100K .......... .......... .......... .......... ..........  1%  516K 18s
   150K .......... .......... .......... .......... ..........  2% 49.2M 14s
   200K .......... .......... ....

In [ ]:
%cd /content/genome-annotation-feature-overlap

!ls -lh data/raw
!ls -lh data/processed

/content/genome-annotation-feature-overlap
total 224M
-rw-r--r-- 1 root root 108M Apr  6 02:50 annotation.gff3
-rw-r--r-- 1 root root 117M Apr  3 06:19 genome.fa
-rw-r--r-- 1 root root  176 Jun 23 05:16 genome.fa.fai
total 19M
-rw-r--r-- 1 root root 9.6M Jun 23 05:16 cds.bed
-rw-r--r-- 1 root root 7.5M Jun 23 05:16 exons.bed
-rw-r--r-- 1 root root 892K Jun 23 05:16 genes.bed
-rw-r--r-- 1 root root   75 Jun 23 05:16 genome_file.txt
-rw-r--r-- 1 root root 892K Jun 23 05:16 promoters_1kb.bed


In [ ]:
!wc -l data/processed/genes.bed
!wc -l data/processed/exons.bed
!wc -l data/processed/cds.bed
!wc -l data/processed/promoters_1kb.bed

27655 data/processed/genes.bed
313952 data/processed/exons.bed
286067 data/processed/cds.bed
27655 data/processed/promoters_1kb.bed


In [ ]:
%%bash

cd /content

# Create project folders
mkdir -p genome-annotation-feature-overlap
cd genome-annotation-feature-overlap
mkdir -p data/raw data/processed results/figures scripts

# Install required tools
apt-get update -qq
apt-get install -y samtools bedtools

# Download raw data
cd data/raw

wget -O annotation.gff3.gz "https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-63/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.63.gff3.gz"

wget -O genome.fa.gz "https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-63/fasta/arabidopsis_thaliana/dna/Arabidopsis_thaliana.TAIR10.dna.toplevel.fa.gz"

gunzip -f annotation.gff3.gz
gunzip -f genome.fa.gz

cd /content/genome-annotation-feature-overlap

# Extract genes
awk 'BEGIN{OFS="\t"}
!/^#/ && $3=="gene" {
  id=".";
  n=split($9,a,";");
  for(i=1;i<=n;i++){
    split(a[i],b,"=");
    if(b[1]=="ID"){
      id=b[2];
    }
  }
  gsub(/^gene:/,"",id);
  print $1,$4-1,$5,id,$6,$7
}' data/raw/annotation.gff3 > data/processed/genes.bed

# Extract exons
awk 'BEGIN{OFS="\t"}
!/^#/ && $3=="exon" {
  id=".";
  n=split($9,a,";");
  for(i=1;i<=n;i++){
    split(a[i],b,"=");
    if(b[1]=="ID"){
      id=b[2];
    }
  }
  gsub(/^exon:/,"",id);
  print $1,$4-1,$5,id,$6,$7
}' data/raw/annotation.gff3 > data/processed/exons.bed

# Extract CDS
awk 'BEGIN{OFS="\t"}
!/^#/ && $3=="CDS" {
  id=".";
  n=split($9,a,";");
  for(i=1;i<=n;i++){
    split(a[i],b,"=");
    if(b[1]=="ID"){
      id=b[2];
    }
  }
  gsub(/^CDS:/,"",id);
  print $1,$4-1,$5,id,$6,$7
}' data/raw/annotation.gff3 > data/processed/cds.bed

# Index genome FASTA
samtools faidx data/raw/genome.fa

# Create genome size file
cut -f1,2 data/raw/genome.fa.fai > data/processed/genome_file.txt

# Create 1 kb promoter regions
bedtools flank \
  -i data/processed/genes.bed \
  -g data/processed/genome_file.txt \
  -l 1000 \
  -r 0 \
  -s > data/processed/promoters_1kb.bed

echo "Recovery completed successfully."

Reading package lists...
Building dependency tree...
Reading state information...
samtools is already the newest version (1.13-4).
bedtools is already the newest version (2.30.0+dfsg-2ubuntu0.1).
0 upgraded, 0 newly installed, 0 to remove and 49 not upgraded.
Recovery completed successfully.


W: Skipping acquire of configured file 'main/source/Sources' as repository 'https://r2u.stat.illinois.edu/ubuntu jammy InRelease' does not seem to provide it (sources.list entry misspelt?)
--2026-06-23 05:20:57--  https://ftp.ensemblgenomes.ebi.ac.uk/pub/plants/release-63/gff3/arabidopsis_thaliana/Arabidopsis_thaliana.TAIR10.63.gff3.gz
Resolving ftp.ensemblgenomes.ebi.ac.uk (ftp.ensemblgenomes.ebi.ac.uk)... 193.62.193.161
Connecting to ftp.ensemblgenomes.ebi.ac.uk (ftp.ensemblgenomes.ebi.ac.uk)|193.62.193.161|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 9571450 (9.1M) [application/x-gzip]
Saving to: ‘annotation.gff3.gz’

     0K .......... .......... .......... .......... ..........  0%  370K 25s
    50K .......... .......... .......... .......... ..........  1%  841K 18s
   100K .......... .......... .......... .......... ..........  1%  517K 18s
   150K .......... .......... .......... .......... ..........  2% 30.8M 13s
   200K .......... .......... ....

In [ ]:
%cd /content/genome-annotation-feature-overlap

!ls -lh data/raw
!ls -lh data/processed

/content/genome-annotation-feature-overlap
total 224M
-rw-r--r-- 1 root root 108M Apr  6 02:50 annotation.gff3
-rw-r--r-- 1 root root 117M Apr  3 06:19 genome.fa
-rw-r--r-- 1 root root  176 Jun 23 05:21 genome.fa.fai
total 19M
-rw-r--r-- 1 root root 9.6M Jun 23 05:21 cds.bed
-rw-r--r-- 1 root root 7.5M Jun 23 05:21 exons.bed
-rw-r--r-- 1 root root 892K Jun 23 05:21 genes.bed
-rw-r--r-- 1 root root   75 Jun 23 05:21 genome_file.txt
-rw-r--r-- 1 root root 892K Jun 23 05:21 promoters_1kb.bed


In [ ]:
!wc -l data/processed/genes.bed
!wc -l data/processed/exons.bed
!wc -l data/processed/cds.bed
!wc -l data/processed/promoters_1kb.bed

27655 data/processed/genes.bed
313952 data/processed/exons.bed
286067 data/processed/cds.bed
27655 data/processed/promoters_1kb.bed


In [ ]:
%cd /content/genome-annotation-feature-overlap
!pwd

/content/genome-annotation-feature-overlap
/content/genome-annotation-feature-overlap


In [ ]:
%%bash
cd /content/genome-annotation-feature-overlap/data/raw

wget -O MYB119_DAPseq_peaks.narrowPeak.gz "https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1925nnn/GSM1925427/suppl/GSM1925427_DAPSeq-MYB_tnt-MYB119_colamp_a-chr1-5_GEM_events.narrowPeak.gz"

--2026-06-23 05:26:02--  https://ftp.ncbi.nlm.nih.gov/geo/samples/GSM1925nnn/GSM1925427/suppl/GSM1925427_DAPSeq-MYB_tnt-MYB119_colamp_a-chr1-5_GEM_events.narrowPeak.gz
Resolving ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)... 130.14.250.10, 130.14.250.11, 2607:f220:41e:250::13, ...
Connecting to ftp.ncbi.nlm.nih.gov (ftp.ncbi.nlm.nih.gov)|130.14.250.10|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 152125 (149K) [application/x-gzip]
Saving to: ‘MYB119_DAPseq_peaks.narrowPeak.gz’

     0K .......... .......... .......... .......... .......... 33%  777K 0s
    50K .......... .......... .......... .......... .......... 67% 71.1M 0s
   100K .......... .......... .......... .......... ........  100% 1.48M=0.1s

2026-06-23 05:26:03 (1.50 MB/s) - ‘MYB119_DAPseq_peaks.narrowPeak.gz’ saved [152125/152125]



In [ ]:
!ls -lh data/raw

total 224M
-rw-r--r-- 1 root root 108M Apr  6 02:50 annotation.gff3
-rw-r--r-- 1 root root 117M Apr  3 06:19 genome.fa
-rw-r--r-- 1 root root  176 Jun 23 05:21 genome.fa.fai
-rw-r--r-- 1 root root 149K Nov  4  2015 MYB119_DAPseq_peaks.narrowPeak.gz


In [ ]:
!gunzip -f data/raw/MYB119_DAPseq_peaks.narrowPeak.gz

In [ ]:
!ls -lh data/raw

total 225M
-rw-r--r-- 1 root root 108M Apr  6 02:50 annotation.gff3
-rw-r--r-- 1 root root 117M Apr  3 06:19 genome.fa
-rw-r--r-- 1 root root  176 Jun 23 05:21 genome.fa.fai
-rw-r--r-- 1 root root 537K Nov  4  2015 MYB119_DAPseq_peaks.narrowPeak


In [ ]:
!head data/raw/MYB119_DAPseq_peaks.narrowPeak

chr5	931194	931395	5:931294	1000	.	2746.0	0.00	999.00	100
chr3	11610429	11610630	3:11610529	983	.	2695.8	0.00	999.00	100
chr3	3503869	3504070	3:3503969	973	.	2666.0	0.00	999.00	100
chr1	1178504	1178705	1:1178604	954	.	2607.0	0.00	999.00	100
chr1	13247604	13247805	1:13247704	891	.	2415.0	0.00	999.00	100
chr2	6298232	6298433	2:6298332	881	.	2385.7	0.00	999.00	100
chr3	13040853	13041054	3:13040953	871	.	2353.3	0.00	999.00	100
chr4	934594	934795	4:934694	856	.	2308.0	0.00	999.00	100
chr5	15029575	15029776	5:15029675	810	.	2166.9	0.00	999.00	100
chr5	5301998	5302199	5:5302098	806	.	2155.0	0.00	999.00	100


In [ ]:
!cut -f1 data/raw/MYB119_DAPseq_peaks.narrowPeak | sort | uniq | head

chr1
chr2
chr3
chr4
chr5


In [ ]:
!cut -f1 data/processed/genes.bed | sort | uniq | head

1
2
3
4
5
Mt
Pt


In [ ]:
%%bash
cd /content/genome-annotation-feature-overlap

awk 'BEGIN{OFS="\t"}
{
  chr=$1;
  gsub(/^Chr/,"",chr);
  gsub(/^chr/,"",chr);
  print chr,$2,$3,"MYB119_peak_"NR,$5,$6
}' data/raw/MYB119_DAPseq_peaks.narrowPeak > data/processed/MYB119_DAPseq_peaks.bed

In [ ]:
!head data/processed/MYB119_DAPseq_peaks.bed

5	931194	931395	MYB119_peak_1	1000	.
3	11610429	11610630	MYB119_peak_2	983	.
3	3503869	3504070	MYB119_peak_3	973	.
1	1178504	1178705	MYB119_peak_4	954	.
1	13247604	13247805	MYB119_peak_5	891	.
2	6298232	6298433	MYB119_peak_6	881	.
3	13040853	13041054	MYB119_peak_7	871	.
4	934594	934795	MYB119_peak_8	856	.
5	15029575	15029776	MYB119_peak_9	810	.
5	5301998	5302199	MYB119_peak_10	806	.


In [ ]:
!wc -l data/processed/MYB119_DAPseq_peaks.bed

9278 data/processed/MYB119_DAPseq_peaks.bed


In [ ]:
!cp data/processed/MYB119_DAPseq_peaks.bed data/processed/candidate_regions.bed

In [ ]:
!wc -l data/processed/candidate_regions.bed
!head data/processed/candidate_regions.bed

9278 data/processed/candidate_regions.bed
5	931194	931395	MYB119_peak_1	1000	.
3	11610429	11610630	MYB119_peak_2	983	.
3	3503869	3504070	MYB119_peak_3	973	.
1	1178504	1178705	MYB119_peak_4	954	.
1	13247604	13247805	MYB119_peak_5	891	.
2	6298232	6298433	MYB119_peak_6	881	.
3	13040853	13041054	MYB119_peak_7	871	.
4	934594	934795	MYB119_peak_8	856	.
5	15029575	15029776	MYB119_peak_9	810	.
5	5301998	5302199	MYB119_peak_10	806	.


In [ ]:
%%bash
cd /content/genome-annotation-feature-overlap

sort -k1,1 -k2,2n data/processed/candidate_regions.bed > data/processed/candidate_regions.sorted.bed
sort -k1,1 -k2,2n data/processed/genes.bed > data/processed/genes.sorted.bed
sort -k1,1 -k2,2n data/processed/exons.bed > data/processed/exons.sorted.bed
sort -k1,1 -k2,2n data/processed/cds.bed > data/processed/cds.sorted.bed
sort -k1,1 -k2,2n data/processed/promoters_1kb.bed > data/processed/promoters_1kb.sorted.bed

In [ ]:
%%bash
cd /content/genome-annotation-feature-overlap

bedtools intersect \
  -a data/processed/candidate_regions.sorted.bed \
  -b data/processed/genes.sorted.bed \
  -wa -wb > results/MYB119_gene_overlaps.tsv

bedtools intersect \
  -a data/processed/candidate_regions.sorted.bed \
  -b data/processed/exons.sorted.bed \
  -wa -wb > results/MYB119_exon_overlaps.tsv

bedtools intersect \
  -a data/processed/candidate_regions.sorted.bed \
  -b data/processed/cds.sorted.bed \
  -wa -wb > results/MYB119_cds_overlaps.tsv

bedtools intersect \
  -a data/processed/candidate_regions.sorted.bed \
  -b data/processed/promoters_1kb.sorted.bed \
  -wa -wb > results/MYB119_promoter_overlaps.tsv

In [ ]:
!wc -l results/MYB119_gene_overlaps.tsv
!wc -l results/MYB119_exon_overlaps.tsv
!wc -l results/MYB119_cds_overlaps.tsv
!wc -l results/MYB119_promoter_overlaps.tsv

5241 results/MYB119_gene_overlaps.tsv
10139 results/MYB119_exon_overlaps.tsv
7139 results/MYB119_cds_overlaps.tsv
3558 results/MYB119_promoter_overlaps.tsv


In [ ]:
%%bash
cd /content/genome-annotation-feature-overlap

total_peaks=$(wc -l < data/processed/candidate_regions.sorted.bed)

gene_overlap=$(bedtools intersect -a data/processed/candidate_regions.sorted.bed -b data/processed/genes.sorted.bed -u | wc -l)
exon_overlap=$(bedtools intersect -a data/processed/candidate_regions.sorted.bed -b data/processed/exons.sorted.bed -u | wc -l)
cds_overlap=$(bedtools intersect -a data/processed/candidate_regions.sorted.bed -b data/processed/cds.sorted.bed -u | wc -l)
promoter_overlap=$(bedtools intersect -a data/processed/candidate_regions.sorted.bed -b data/processed/promoters_1kb.sorted.bed -u | wc -l)

echo -e "feature_type\toverlapping_peaks\ttotal_peaks\tpercentage" > results/MYB119_overlap_summary.tsv

echo -e "genes\t$gene_overlap\t$total_peaks\t$(awk -v a=$gene_overlap -v b=$total_peaks 'BEGIN{printf "%.2f", (a/b)*100}')" >> results/MYB119_overlap_summary.tsv

echo -e "exons\t$exon_overlap\t$total_peaks\t$(awk -v a=$exon_overlap -v b=$total_peaks 'BEGIN{printf "%.2f", (a/b)*100}')" >> results/MYB119_overlap_summary.tsv

echo -e "CDS\t$cds_overlap\t$total_peaks\t$(awk -v a=$cds_overlap -v b=$total_peaks 'BEGIN{printf "%.2f", (a/b)*100}')" >> results/MYB119_overlap_summary.tsv

echo -e "promoters_1kb\t$promoter_overlap\t$total_peaks\t$(awk -v a=$promoter_overlap -v b=$total_peaks 'BEGIN{printf "%.2f", (a/b)*100}')" >> results/MYB119_overlap_summary.tsv

In [ ]:
!cat results/MYB119_overlap_summary.tsv

feature_type	overlapping_peaks	total_peaks	percentage
genes	5054	9278	54.47
exons	4848	9278	52.25
CDS	3419	9278	36.85
promoters_1kb	3197	9278	34.46
